File: 4164-last-remaining-integer-after-alternating-deletion-operations/4164-last-remaining-integer-after-alternating-deletion-operations.py
class Solution:
    def lastInteger(self, n: int) -> int:
        start, end , step = 1, n , 1
        while start!=end:
            n = abs(start-end)//abs(step)+1
            if (n-1).bit_count()==1 and n.bit_length()&1: 
                return start
            if abs(start-end) //abs(step) & 1:
                start, end, step = end-step,start, -2*step
            else:
                start, end, step = end, start, -2 *step
        return start

File: 4164-last-remaining-integer-after-alternating-deletion-operations/4164-last-remaining-integer-after-alternating-deletion-operations.cpp
class Solution {
public:
    long long lastInteger(long long n) {
        long long start = 1, end = n, step = 1;
        unsigned long long m;
        for (m = (unsigned long long) n; m != (1 | (1LL << (bit_width(m) - 1))); m -= m >> 1)
            tie(start, end, step) = make_tuple(end - ((m & 1) ? 0: step), start, -2 * step);
            
        return bit_width(m) & 1 ? start : end;
    }
};